File: go/construct-binary-search-tree-from-preorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNodeBound struct {
	Node  *TreeNode
	Bound int
}

func bstFromPreorder(preorder []int) *TreeNode {
	root := TreeNode{Val: preorder[0]}
	s := make([]TreeNodeBound, 0)
	s = append(s, TreeNodeBound{Node: &root, Bound: 1001})
	for i := 1; i < len(preorder); i++ {
		node := TreeNode{Val: preorder[i]}
		peak := len(s) - 1
		for s[peak].Bound < node.Val {
			s = s[:peak]
			peak -= 1
		}
		parent := s[peak].Node
		bound := s[peak].Bound
		if node.Val < parent.Val {
			parent.Left = &node
			s = append(s, TreeNodeBound{Node: &node, Bound: parent.Val})
		} else {
			parent.Right = &node
			s = append(s, TreeNodeBound{Node: &node, Bound: bound})
		}
	}
	return &root
}

File: python/construct-binary-search-tree-from-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        root = TreeNode(preorder[0])
        s = [(root, 1001)]
        for i in range(1, len(preorder)):
            node = TreeNode(preorder[i])
            while s[-1][1] < node.val:
                s.pop()
            parent, upper_bound = s[-1]
            if node.val < parent.val:
                parent.left = node
                s.append((node, parent.val))
            else:
                parent.right = node
                s.append((node, upper_bound))
        return root
